### DISCUSSION TITLE
Apple authentication with Supabase in Flutter: Error 400: invalid_request

### DISCUSSION THREAD

**Discussion by Alexandru-C** [supabase/supabase#21297] [Questions] [answered]
I am trying to implement Apple authentication with Supabase in my Flutter app. I followed all the steps in the Supabase documentation, but when I try to log in I get the following error: Unhandled Exception: AuthException(message: Unacceptable audience in id_token, statusCode: 400)

The stacktrace is:
0      GotrueFetch.request (package:gotrue/src/fetch.dart:99:7)
<asynchronous suspension>
1      GoTrueClient.signInWithIdToken (package:gotrue/src/gotrue_client.dart:368:22)
<asynchronous suspension>
2      AuthRepositoryImpl.signInWithApple (package:onboarding_app_data/repostories_impl/auth_repository_impl.dart:98:22)
<asynchronous suspension>
3      AuthRepositoryImpl.authenticateWithProvider (package:onboarding_app_data/repostories_impl/auth_repository_impl.dart:39:18)
<asynchronous suspension>
4      AuthenticationViewBloc._authenticateEventHandler (package:onboarding_app/authentication/authentication_view_bloc.dart:53:9)
<asynchronous suspension>
5      Bloc.on.<anonymous closure>.handleEvent (package:bloc/src/bloc.dart:229:13)

The apple auth works from an Android device using 
`await supabaseInstance.client.auth.signInWithOAuth(
        OAuthProvider.apple,
        redirectTo: 'MY_CALLBACK');`

But on an IOS device I get the issues described above.
    
Besides the code provided here: https://supabase.com/docs/guides/auth/social-login/auth-apple?platform=flutter  I added the crypto package to manage the SHA hashing. 
    
   Did anyone mange to make this work?

**Comment by Alexandru-C** [comment]
Kind of fixed this by putting the Bundle ID instead of service id in SOCIAL_EXTERNAL_APPLE_CLIENT_ID field and works for users previously logged in, but for new users I get Bad ID token, statusCode: 400

**Comment by dshukertjr** [comment, marked as answer]
For Apple sign-in on iOS, what you want to do is to enter the bundle ID on `Authorized Client IDs` section of the Supabase dashboard on Apple provider page. 

For Android, you want to add the service ID and signing key into your Supabase dashboard on Apple provider page.

**Reply by kanapanda** [reply]
Is there a way to set the 'Authorized Client IDs' for local development ?
My setup works fine in production (dashboard) but cannot replicate it locally: I get the 'Unacceptable audience in id_token' error.
I tried to set the authorized IDs in config.toml but could not find any info in the documentation.

**Reply by dshukertjr** [reply]
You can provide it as the service ID. If you already have a service ID, you can just provide the additional client IDs as comma separated string.

**Reply by QuDaMyker** [reply]
Also have this error "AuthException(message: Unacceptable audience in id_token, statusCode: 400)", although idToken and accessToken are true

```javascript
Future<Either<FailModel, AuthResponse>> nativeGoogleSignIn() async {
    try {
      final webClientId = '${dotenv.get('google_client_id')}';
      print(webClientId);
      const iosClientId = 'my-ios.apps.googleusercontent.com';
      final GoogleSignIn googleSignIn = GoogleSignIn(
        clientId: iosClientId,
        serverClientId: webClientId,
      );
      final googleUser = await googleSignIn.signIn();
      final googleAuth = await googleUser!.authentication;
      final accessToken = googleAuth.accessToken;
      final idToken = googleAuth.idToken;

      print(accessToken);
      print(idToken);

      if (accessToken == null) {
        throw 'No Access Token found.';
      }
      if (idToken == null) {
        throw 'No ID Token found.';
      }

      var res = await supabase.auth.signInWithIdToken(
        provider: OAuthProvider.google,
        idToken: idToken,
        accessToken: accessToken,
      );

      return Right(res);
    } catch (e) {
      debugPrint(e.toString());
      return Left(FailModel(message: e.toString()));
    }
  }
  ```

**Reply by dshukertjr** [reply]
`Unacceptable audience in id_token` means that you don't have the client ID used for the authentication set in the Supabase dashboard. Double-check what you have and what you are using in your Flutter code.

**Reply by QuDaMyker** [reply]
> `Unacceptable audience in id_token` means that you don't have the client ID used for the authentication set in the Supabase dashboard. Double-check what you have and what you are using in your Flutter code.

Thanks you I was fixed, 
But I meet order bugs, can you help me review then give me some methods to fix it ? thank you so much

https://github.com/orgs/supabase/discussions/26791

**Reply by RemyNtshaykolo** [reply]
Don't forget to supabase stop and supabase start after changes in the .toml file

**Comment by Lazizbek97** [comment]
for ios, I put `bundleid,ClientID`
then it worked

**Reply by rinathama** [reply]
what legend you are. Thank you
